controllers/kubeadmconfig_controller.go

```diff
@@ -140,10 +128,39 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, err
 	}
 
+	switch {
 	// Wait patiently for the infrastructure to be ready
-	if !cluster.Status.InfrastructureReady {
+	case !cluster.Status.InfrastructureReady:
 		log.Info("Infrastructure is not ready, waiting until ready.")
 		return ctrl.Result{}, nil
+	// bail super early if it's already ready
+	case config.Status.Ready && machine.Status.InfrastructureReady:
+		log.Info("ignoring config for an already ready machine")
+		return ctrl.Result{}, nil
+	// Ignore machines that already have bootstrap data we didn't generate
+	case machine.Spec.Bootstrap.Data != nil && !config.Status.Ready:
+		// TODO: mark the config as ready?
```

Comment:
+1 to marking the Config as ready in this case. This will also help with rebuilding Status if lost.

Reply:
If this provider wasn't used to generate the user data, we shouldn't set it as ready. There is only one case I can think of that this is appropriate: if the bootstrap config reference points to this kubeadm config object and the bootstrap.data field is already populated, which covers to loss of Status

Reply:
If we have an ownerRef to a Machine, I'd assume that it would be because the Machine Controller set the ownerRef based on the bootstrap Config Reference, so us getting to this point should already mean that is the case.

Reply:
I was more replying to the comment above the `case` condition, which seems misleading. +1 to what @detiber pointed out